pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -0,0 +1,419 @@
+	if err != nil {
+		klog.Errorf("[memorymanager] could not initialize checkpoint manager: %v, please drain node and remove policy state file", err)
+		return err
+	}
+	m.state = stateImpl
+
+	err = m.policy.Start(m.state)
+	if err != nil {
+		klog.Errorf("[memorymanager] policy start error: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// AddContainer saves the value of requested memory for the guaranteed pod under the state and set memory affinity according to the topolgy manager
+func (m *manager) AddContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
+	m.Lock()
+	m.containerMap.Add(string(pod.UID), container.Name, containerID)
+	m.Unlock()
+
+	// Get NUMA node affinity of blocks assigned to the container during Allocate()
+	var nodes []string
+	for _, block := range m.state.GetMemoryBlocks(string(pod.UID), container.Name) {
+		for _, nodeID := range block.NUMAAffinity {
+			nodes = append(nodes, strconv.Itoa(nodeID))
+		}
+	}
+
+	if len(nodes) < 1 {
+		klog.V(5).Infof("[memorymanager] update container resources is skipped due to memory blocks are empty")
+		return nil
+	}
+
+	affinity := strings.Join(nodes, ",")
+	klog.Infof("[memorymanager] Set container %q cpuset.mems to %q", containerID, affinity)
+	err := m.containerRuntime.UpdateContainerResources(containerID, &runtimeapi.LinuxContainerResources{CpusetMems: affinity})
+	if err != nil {
+		klog.Errorf("[memorymanager] AddContainer error: error updating cpuset.mems for container (pod: %s, container: %s, container id: %s, err: %v)", pod.Name, container.Name, containerID, err)
+
+		m.Lock()
+		err = m.policyRemoveContainerByRef(string(pod.UID), container.Name)
+		if err != nil {
+			klog.Errorf("[memorymanager] AddContainer rollback state error: %v", err)
+		}
+		m.Unlock()
+	}
+	return err
+}
+
+// Allocate is called to pre-allocate memory resources during Pod admission.
+func (m *manager) Allocate(pod *v1.Pod, container *v1.Container) error {
+	// Garbage collect any stranded resources before allocation
+	m.removeStaleState()
+
+	m.Lock()
+	defer m.Unlock()
+
+	// Call down into the policy to assign this container memory if required.
+	if err := m.policy.Allocate(m.state, pod, container); err != nil {
+		klog.Errorf("[memorymanager] Allocate error: %v", err)
+		return err
+	}
+	return nil
+}
+
+// RemoveContainer removes the container from the state
+func (m *manager) RemoveContainer(containerID string) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// if error appears it means container entry already does not exist under the container map
+	podUID, containerName, err := m.containerMap.GetContainerRef(containerID)
+	if err != nil {
+		klog.Warningf("[memorymanager] Failed to get container %s from container map error: %v", containerID, err)
+		return nil
+	}
+
+	err = m.policyRemoveContainerByRef(podUID, containerName)
+	if err != nil {
+		klog.Errorf("[memorymanager] RemoveContainer error: %v", err)
+		return err
+	}
```

Comment:
I think you can just call `m.policyRemoveContainerById()` here if you want.
Unless you explicitly wanted to print out that warning above.

If so, I would still do this instead (though I don't think it's necessary):
```
	// if error appears it means container entry already does not exist under the container map
	_, _, err := m.containerMap.GetContainerRef(containerID)
	if err != nil {
		klog.Warningf("[memorymanager] Failed to get container %s from container map error: %v", containerID, err)
		return nil
	}

	err = m.policyRemoveContainerByRef(containerId)
	if err != nil {
		klog.Errorf("[memorymanager] RemoveContainer error: %v", err)
		return err
	}
```

Reply:
Hm, I do not have such a method under the manager, I can add it, but I unsure if it really needed.

Reply:
I see. Yeah, there is one in the `CPUManager` that basically does the same thing you are doing here (but without the log message). I would probably add it for symmetry with the `CPUManager`, but it's up to you since we plan to reabstract this together with the `CPUManager` later.